### PR TITLE
Support backward of backward for Relu and add a new gradient checker by comparing theoretical and numerical Jacobian.

### PR DESCRIPTION
**Pull request by qingqing01**
1. Use Relu to get through the whole process from C++ to Python.
   - Add double backward  op for Relu (named `relu_grad_grad` op).
   - Add register for `relu_grad_grad`.

Add `ReluDoubleGradMaker` when register `relu_grad`:

```c++
REGISTER_OPERATOR(relu_grad, ops::ActivationOpGrad,
                  paddle::framework::SingleOpInplaceInToOut,
                  ops::ReluDoubleGradMaker);
```

2. Add a new gradient checker by comparing theoretical and numerical Jacobian. 
   - support to check program, not one op.
   - support to check gradients by API `grad_check`.
   - support to check double gradients by API `double_grad_check`.

3. Add gradients testing  for `relu_grad_grad`.
 
4. `fluid.backward.calc_gradient` is used to  append the backward pass,  which is similar interface to `tf.gradients`.

For example:

```python
import paddle.fluid as pd
from paddle.fluid.backward import calc_gradient

# relu
x = pd.layers.data(name='x', shape=[2,4], dtype='float32', append_batch_size=False)
x.stop_gradient=False
y = pd.layers.relu(x)
z = calc_gradient([y], x)
c = calc_gradient(z, x)

print(pd.default_main_program())
```

Note,  this PR doest not use  macro to register double grad for activations since only add one for relu.  If adding more in the future, will use  macro for more general.

### CODE REVIEW

File: paddle/fluid/operators/activation_op.cc [hunk 1]
```diff
@@ -586,10 +586,60 @@ REGISTER_ACTIVATION_OP_MAKER(Log, LogDoc);
...
+    auto* op = new ::paddle::framework::OpDesc();
+    op->SetType("relu_grad_grad");
+    // input1: Out
+    op->SetInput("Out", Input("Out"));
+    // X@GRAD@GRAD: ddx
+    op->SetInput(framework::DoubleGradVarName("X"),
+                 OutputGrad(framework::GradVarName("X")));
+    op->SetAttrMap(Attrs());
+    // Out@GRAD@GRAD: ddy
+    op->SetOutput(framework::GradVarName("Out"), InputGrad("Out"));
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/activation_op.cc]: `// ReluGradGrad: ddy = ddx if y >= 0 else 0`

这个公式看起来和dy没什么关系，所以你确定这里用把dy放outputs么？

Comment (qingqing01) [reply paddle/fluid/operators/activation_op.cc]: Yeah, y is the input of relu_grad, relu_grad_grad also needs to calculate dy.

File: python/paddle/fluid/tests/unittests/gradient_checker.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+from itertools import product
+
+import paddle.fluid as fluid
+import paddle.fluid.core as core
+from paddle.fluid.executor import Executor
+from paddle.fluid.backward import calc_gradient
+from paddle.fluid.backward import _append_grad_suffix_, _as_list
+
+
+def _product(t):
```
Comment (wanghaoshuang) [inline python/paddle/fluid/tests/unittests/gradient_checker.py:30]: 为什么不用np.product呢？

File: python/paddle/fluid/tests/unittests/gradient_checker.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+               eps=1e-6,
+               atol=1e-5,
+               rtol=1e-3,
+               raise_exception=True):
+    """
+    Check numerical and analytical gradients for dy/dx.
+    Each Jacobian gradients is a 2-D array with shape [xi_size, yi_size].
+
+    Args:
+        program (Program): a Program with forward pass.
```
Comment (wanghaoshuang) [inline python/paddle/fluid/tests/unittests/gradient_checker.py]: 这行应该删掉，和下面一行重复了。

File: python/paddle/fluid/tests/unittests/test_nn_grad.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+class TestReluDoubleGradCheck(TestGradCheck):
+    @prog_scope()
+    def func(self, place):
+        # the shape of input variable shoule be clearly specified, not inlcude -1.
+        shape = [2, 8]
+        eps = 0.005
+        dtype = np.float64
+
+        #x = layers.data('x', shape, False, dtype)
```
Comment (wanghaoshuang) [inline python/paddle/fluid/tests/unittests/test_nn_grad.py]: 这个也可以删掉了吧？

**Review by wanghaoshuang — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/gradient_checker.py (hunk 3 above)
Comment (qingqing01) [reply python/paddle/fluid/tests/unittests/gradient_checker.py]: Done. Thanks!

File: python/paddle/fluid/tests/unittests/test_nn_grad.py (hunk 4 above)
Comment (qingqing01) [reply python/paddle/fluid/tests/unittests/test_nn_grad.py]: Done.

File: python/paddle/fluid/tests/unittests/gradient_checker.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/fluid/tests/unittests/gradient_checker.py:30]: Done, Thanks!

**Comment by qingqing01** [comment]
@sneaxiy I have deleted `DoubleGradVarName`.
@wanghaoshuang Update code according your comments.

Thanks!

**Review by wanghaoshuang — DISMISSED** [review]

**Comment by qingqing01** [comment]
```
[16:03:43]W:	 [Step 1/1] + API_CHANGE=python/paddle/fluid/tests/unittests/CMakeLists.txt
[16:03:43]W:	 [Step 1/1] + echo 'checking CMakeLists.txt change, PR: 16862, changes: python/paddle/fluid/tests/unittests/CMakeLists.txt'
[16:03:43]W:	 [Step 1/1] + '[' python/paddle/fluid/tests/unittests/CMakeLists.txt ']'
[16:03:43]W:	 [Step 1/1] + '[' 16862 '!=' '' ']'
[16:03:43]W:	 [Step 1/1] + '[' CMakeLists.txt == paddle/fluid/API.spec ']'
[16:03:43]W:	 [Step 1/1] + '[' CMakeLists.txt == CMakeLists.txt ']'
[16:03:43] :	 [Step 1/1] checking CMakeLists.txt change, PR: 16862, changes: python/paddle/fluid/tests/unittests/CMakeLists.txt
[16:03:43]W:	 [Step 1/1] ++ python /paddle/tools/check_pr_approval.py 1 6836917 46782768 30176695
[16:03:43]W:	 [Step 1/1] ++ curl -H 'Authorization: token 5ae37427a95d9e51fb0ea3c71121b75d5c2ebefa' 'https://api.github.com/repos/PaddlePaddle/Paddle/pulls/16862/reviews?per_page=10000'
[16:03:43]W:	 [Step 1/1]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[16:03:43]W:	 [Step 1/1]                                  Dload  Upload   Total   Spent    Left  Speed
[16:03:43]W:	 [Step 1/1] 
[16:03:44]W:	 [Step 1/1]   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[16:03:45]W:	 [Step 1/1]   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[16:03:45]W:	 [Step 1/1]   1  9156    1   125    0     0     64      0  0:02:23  0:00:01  0:02:22    64
[16:03:45]W:	 [Step 1/1] 100  9156  100  9156    0     0   4742      0  0:00:01  0:00:01 --:--:--  4744
[16:03:45] :	 [Step 1/1] current pr 16862 got approvals: FALSE
[16:03:45] :	 [Step 1/1] You must have one RD (luotao1 or chengduoZH or XiaoguangHu01) approval for the cmakelist change! CMakeLists.txt
[16:03:45]W:	 [Step 1/1] + APPROVALS=FALSE
```

The unit testing failed since Python on Windows CI have a same python package. Rename decorators to decorator_helper. So change the `python/paddle/fluid/tests/unittests/CMakeLists.txt`.


Please help to review. Thanks!  @chengduoZH @luotao1 @XiaoguangHu01

**Review by luotao1 — APPROVED** [review]
LGTM for cmakelist.txt

**Comment by luotao1** [comment]
@tianshuo78520a It seems #16944 don't fix this problem https://github.com/PaddlePaddle/Paddle/pull/16862#issuecomment-485330558

**Pull request merged**